### ISSUE TITLE
Set Number of Trials based on Dynamic Hyperparameters (Dynamic Trials?)

### ISSUE THREAD

**Issue by koh-joshua** [optuna/optuna#1611]
I have two sets of hyperparameters:
set1: master_param1
set2: slave_param1, slave_param2
Set2 (slave_param1, slave_param2) parameters are dependent on master_param1. 

I want to fix a certain number of trials for set2 (slave_param1, slave_param2) based on master_param1. Example:

def objective(trial):
       master_param1 = trial.suggest_int('master_param1', 0, 100)
       for n in range(100):   # set 100 trials based on master_param1, in this case 100 trials
                  slave_param1 = trial.suggest_int('slave_param1', 0, 10)
                  slave_param2 = trial.suggest_int('slave_param2', 0, 50)
                  dummy_model = model(master_param1, slave_param1, slave_param2)
                  score = dummy_model.evaluate(x, y)
                  return score

Basically I want to set number of trials for a second set of hyperparameters based on the first set of hyperparameters suggested by Optuna. I am not sure whether this makes sense?

**Comment by HideakiImamura** [comment]
Thanks for the question! First, I would like to know how your objective function return the value. Your objective is like this? That is, `return score` statement is out of the for loop?
```python
def objective(trial):
master_param1 = trial.suggest_int('master_param1', 0, 100)
for n in range(100): # set 100 trials based on master_param1, in this case 100 trials
    slave_param1 = trial.suggest_int('slave_param1', 0, 10)
    slave_param2 = trial.suggest_int('slave_param2', 0, 50)
    dummy_model = model(master_param1, slave_param1, slave_param2)
    score = dummy_model.evaluate(x, y)
return score
```

**Comment by koh-joshua** [comment]
Thanks for the reply. I don't have an actual objective function, it's just pseudo-code at the moment.
I am not sure how to set up the objective function with dynamic trials. 

Perhaps one possible way is to have nested study for master_param1?:

def master_objective(trial):
    master_param1 = trial.suggest_int('master_param1', 0, 100)
    def slave_objective(trial):
        slave_param1 = trial.suggest_int('slave_param1', 0, 10)
        slave_param2 = trial.suggest_int('slave_param2', 0, 50)
        dummy_model = model(master_param1, slave_param1, slave_param2)
        score = dummy_model.evaluate(x, y)
        return score
    nested_study = optuna.create_study()
    nested_study.optimize(slave_objective, n_trials=100, n_jobs=-1)
    best_score = nested_study.best_value
    return best_score

master_study = optuna.create_study()
master_study.optimize(master_objective, n_trials=500, n_jobs=-1)

**Comment by HideakiImamura** [comment]
Thanks for the reply! I understood what you want to do. It's very interesting to me. We (Optuna core-dev team) did not consider nesting study objects, but I think it can work. I have no idea about a concrete objective example, but if you have in the future, could you make an example? I would like to know the specific use-case.

By the way, if we use `study.Study.enqueue_trial`, we can already define an objective that has the fixed master parameter with a certain number.
```python
def objective(trial):
    master_param1 = trial.suggest_int('master_param', 0, 100)
    slave_param1 = trial.suggest_int('slave_param1', 0, 10)
    slave_param2 = trial.suggest_int('slave_param2', 0, 50)
    dummy_model = model(master_param, slave_param1, slave_param2)
    score = dummy_model.evaluate(x, y)
    return score

study = optuna.create_study()
for i in range(500):
    study.optimize(objective, n_trials=1)
    master_param = study.get_trials[-1].params['master_param']
    for _ in range(100):
        study.enqueue_trial({'master_param': master_param})
    study.optimize(objective, n_trials=100)
```

**Comment by koh-joshua** [comment]
Thank you for the suggestions! 

Got an error relating to **study.get_trials**:
Traceback (most recent call last):
  File "C:\Anaconda3\envs\HPO\lib\site-packages\IPython\core\interactiveshell.py", line 3417, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-17-999b87c8bb4d>", line 1, in <module>
    master_param = study.get_trials[-1].params['master_param']
TypeError: 'method' object is not subscriptable

However, using **study.trials[-1]**.params['master_param'] worked.

**Comment by koh-joshua** [comment]
I am closing this post as all questions have been addressed. Thank you, Team  Optuna.

**Comment by koh-joshua** [comment]
Just want to drop a Thank You note to the team over at Optuna. I was able to use Optuna to develop a custom AutoML solution to address a challenging 50-year-old puzzle in remote sensing. The system got published in a prestigious plant journal recently. I used TPE as the main engine to drive model selection, optimization and evaluation, taking inspiration from existing AutoML systems as an example. My engagement with Optuna remains unique as it happened during COVID 2020. Thank you once again.
